File: 백준/Gold/2696. 중앙값 구하기/중앙값 구하기.py
import sys
from math import ceil
input = lambda : sys.stdin.readline().rstrip()
from heapq import heappop, heappush

tc = int(input())
for _ in range(tc):
    n = int(input())
    left,mid,right=[],[],[]
    index=0
    result=[]
    print(n//2+1)
    nums = []
    for i in range(ceil(n/10)):
        nums+=list(map(int,input().split()))
    for i in nums:
        if not mid:
            mid.append(i)
        else:
            if i>=mid[0]:
                heappush(right,i)
            else:
                heappush(left,-i)
            if len(left)< len(right):
                heappush(left, -mid.pop())
                mid.append(heappop(right))
            elif len(left) > len(right):
                heappush(right,mid.pop())
                mid.append(-heappop(left))

        if index%2==0:
            # print(left,mid,right)
            result.append(mid[0])
        if len(result)==10:
            print(*result)
            result=[]
        index+=1

    if result:
        print(*result)

File: 백준/Gold/2696. 중앙값 구하기/중앙값 구하기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int tc = Integer.parseInt(br.readLine());
        while (tc-- > 0) {
            int n = Integer.parseInt(br.readLine());
            int[] nums = new int[n];
            int idx = 0;
            while (idx < n) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                while (st.hasMoreTokens()) {
                    nums[idx++] = Integer.parseInt(st.nextToken());
                }
            }

            PriorityQueue<Integer> left = new PriorityQueue<>(Collections.reverseOrder());
            PriorityQueue<Integer> right = new PriorityQueue<>();
            List<Integer> result = new ArrayList<>();

            int mid = nums[0];
            result.add(mid);
            for (int i = 1; i < n; i++) {
                int num = nums[i];
                if (num >= mid) {
                    right.offer(num);
                } else {
                    left.offer(num);
                }

                if (left.size() < right.size()) {
                    left.offer(mid);
                    mid = right.poll();
                } else if (left.size() > right.size()) {
                    right.offer(mid);
                    mid = left.poll();
                }

                if (i % 2 == 0) {
                    result.add(mid);
                }
            }

            sb.append(result.size()).append("\n");
            for (int i = 0; i < result.size(); i++) {
                sb.append(result.get(i)).append(" ");
                if ((i + 1) % 10 == 0) sb.append("\n");
            }
            if (result.size() % 10 != 0) sb.append("\n");
        }

        System.out.print(sb);
    }
}
